internal/opa_inspect/opa_inspect.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// Functions for inspecting rego code and annotations
+package opa_inspect
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/spf13/afero"
+)
+
+func inspectSingle(path, module string) ([]*ast.AnnotationsRef, error) {
+	mod, err := ast.ParseModuleWithOpts(path, module, ast.ParserOptions{ProcessAnnotation: true})
+	if err != nil {
+		return nil, err
+	}
+
+	as, errs := ast.BuildAnnotationSet([]*ast.Module{mod})
+	if len(errs) > 0 {
+		return nil, errors.New(errs.Error())
+	}
+
+	results := make([]*ast.AnnotationsRef, 0, len(mod.Rules))
+	for _, rule := range mod.Rules {
+		results = append(results, as.Chain(rule)...)
+	}
+
+	return results, nil
+}
+
+func inspectMultiple(paths, modules []string) ([]*ast.AnnotationsRef, error) {
+	numPaths := len(paths)
+	numModules := len(modules)
+	if numPaths != numModules {
+		return nil, fmt.Errorf("mismatched number of paths and modules: %d != %d", numPaths, numModules)
+	}
+
+	results := make([]*ast.AnnotationsRef, 0, numPaths)
+	for i := 0; i < numPaths; i++ {
+		r, err := inspectSingle(paths[i], modules[i])
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results, r...)
+	}
+
+	return results, nil
+}
+
+func isWarnOrDeny(rule *ast.AnnotationsRef) bool {
+	r := rule.GetRule()
+	if r == nil {
+		return false
+	}
+	ruleName := r.Head.Name.String()
+	return ruleName == "deny" || ruleName == "warn"
+}
+
+func hasAnnotations(rule *ast.AnnotationsRef) bool {
+	return rule.Annotations != nil
+}
+
+// Interesting rules are warns, denies, and anything with an annotation,
+// which includes some package scoped annotations not assocated with a rule
+func interestingRulesOnly(results []*ast.AnnotationsRef) ([]*ast.AnnotationsRef, error) {
+	filteredResults := make([]*ast.AnnotationsRef, 0, len(results))
+	for _, rule := range results {
+		if isWarnOrDeny(rule) || hasAnnotations(rule) {
```

Comment:
It would be useful to document this behavior in a more "public" place. I did not expect rules without annotations to be excluded but I agree that it does make sense to do so.

Reply:
The presence of the --raw flag did provide some clue about it before.

But note: Warns/denies without annotations would show in the json. And in the very latest revision (a short time ago), the also now appear in the text output. The things I'm mainly excluding are the `lib.useful_thing` and the `_some_helper` type of rules.